File: 0301-0400/0336-Palindrome Pairs/0336-Palindrome Pairs.cpp
class Solution {
public:
    vector<vector<int>> palindromePairs(vector<string>& words) {
        vector<vector<int>> result;
        unordered_map<string, int> table;
        for (int i = 0; i < words.size(); ++i) {
            string word = words[i];
            reverse(word.begin(), word.end());
            table[word] = i;
        }
        
                
        for (int i = 0; i < words.size(); ++i) {
            for (int j = 1; j <= words[i].size(); ++j) {
                string left = words[i].substr(0, j);
                string right = words[i].substr(j);
                if (isPalidrome(right) && table.count(left) && table[left] != i) {
                    result.push_back({i, table[left]});
                }
                if (isPalidrome(left) && table.count(right) && table[right] != i) {
                    result.push_back({table[right], i});
                }
            }
        }
        
        if (table.count("")) {
            for (int i = 0; i < words.size(); ++i) {
                if (words[i] != "" && isPalidrome(words[i])) {
                    result.push_back({i, table[""]});
                }
            }
        }
        
        return result;
    }
    
private:
    bool isPalidrome(string s) {
        if (s.size() == 0) {
            return true;
        }
        int start = 0, end = s.size() - 1;
        while (start < end) {
            if (s[start] != s[end]) {
                return false;
            }
            ++start;
            --end;
        }
        return true;
    }
};

File: 0301-0400/0336-Palindrome Pairs/0336-Palindrome Pairs.py
class Solution:
    def palindromePairs(self, words: List[str]) -> List[List[int]]:
        result = []
        table = {word[::-1]: i for i, word in enumerate(words)}
        for i, word in enumerate(words):
            for j in range(1, len(word) + 1):
                left = word[:j]
                right = word[j:]
                if left == left[::-1] and right in table and table[right] != i:
                    result.append([table[right], i])
                if right == right[::-1] and left in table and table[left] != i:
                    result.append([i, table[left]])
                    
        if '' in table:
            index = table['']
            for i, word in enumerate(words):
                if i != index and word == word[::-1]:
                    result.append([i, index])
        return result
